### PR TITLE
[BEAM-11105] Basic Watermark Estimation (Wall Clock Observing)

### PR DESCRIPTION
**Pull request by damccorm**
**Summary of Changes**

This pr is the first step towards full watermark estimation support - it introduces the mechanism used for watermark estimation including:

- Validation (graph/fn.go)
- Code generation (genx/genx.go)
- Execution (exec/sdf.go and exec/sdf_invokers.go)
- Propogating result back to the runner (exec/datasource.go, harness.go)
- A concrete implementation of a wall clock observing estimator (sdf/watermarkEstimator.go)

This is built according to the design specified here - https://docs.google.com/document/d/1DqCYJ-J1YGNelCRIcN5v6BQsZxJB2l5uWBmN4ti--Ew/edit?usp=sharing

**Next Steps**

This work does not attempt any of the following (though it is built in such a way that those should be cleanly supported in subsequent PRs):

- Watermark state
- Manual estimators (passed into ProcessElement)
- Timestamp observing estimators (call ObserveTimestamp on each emitted timestamp)

**Manual Testing**

On top of all the unit tests introduced, I hacked the existing textio sdf doFn and added the following function:

```
func (fn *readSdfFn) CreateWatermarkEstimator() sdf.WallTimeWatermarkEstimator {
	return sdf.WallTimeWatermarkEstimator{}
}
```

and ran wordCount (with the textio sdf version) on dataflow so that I could observe the output watermark was getting correctly updated to the current wall clock time:

image

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: sdks/go/pkg/beam/core/runtime/exec/plan.go
Comment (youngoli) [inline sdks/go/pkg/beam/core/runtime/exec/plan.go:27]: Is timestamppb the type we already represent watermarks as internally? From a cursory look it feels weird to represent it with a protobuf type, but I'm also not sure what we currently use so I just want to check.

File: sdks/go/pkg/beam/core/runtime/exec/sdf.go
Comment (youngoli) [inline sdks/go/pkg/beam/core/runtime/exec/sdf.go:342]: Creating a watermark estimator doesn't need to be done for each element, does it? If not, this should go in StartBundle.

File: sdks/go/pkg/beam/core/runtime/exec/sdf.go
Comment (youngoli) [inline sdks/go/pkg/beam/core/runtime/exec/sdf.go:696]: Maybe it's because I'm not well-versed in how watermarks work, but why store output watermarks in a map like this if we set the same value for every single output? Why not just store the single timestamp? Is it because of some upcoming work that's going to build on this?

**Review by youngoli — COMMENTED** [review]
For the most part this looks really good. I have a few comments/questions though, and it might be beneficial to get approval from someone who understands watermarks better than I do, but as far as the Go SDK details are concerned, this is looking good.

File: sdks/go/pkg/beam/core/graph/fn.go
Comment (damccorm) [reply sdks/go/pkg/beam/core/graph/fn.go]: Whoops, good catch

File: sdks/go/pkg/beam/core/graph/fn.go
Comment (damccorm) [reply sdks/go/pkg/beam/core/graph/fn.go:968]: Yeah, there will definitely be more added in the future (I actually already have those changes locally 😄 )

File: sdks/go/pkg/beam/core/graph/fn_test.go
Comment (damccorm) [reply sdks/go/pkg/beam/core/graph/fn_test.go:274]: I actually like leaving it for 2 reasons:

1) (more important) It will be relevant as these tests are expanded to future use cases
2) (probably a weaker argument) This is still part of the testable interface - even if we know that under the covers we don't make use of KV info for watermark estimation testing, there's not a reason that _must_ be true. This test protects us if we want to make implementation changes that somehow involve KV awareness (like my first point)

File: sdks/go/pkg/beam/core/runtime/exec/plan.go
Comment (damccorm) [reply sdks/go/pkg/beam/core/runtime/exec/plan.go:27]: It is in several places, but most importantly in our OutputWatermarks proto - https://github.com/apache/beam/blob/81d3de2007f82703d93c420bf03b00d00c03e96e/sdks/go/pkg/beam/model/fnexecution_v1/beam_fn_api.pb.go#L922

File: sdks/go/pkg/beam/core/runtime/exec/sdf.go
Comment (damccorm) [reply sdks/go/pkg/beam/core/runtime/exec/sdf.go:342]: It actually does - watermark estimation is done at the element/restriction level. This will become more important when we're taking in watermark state which is created while giving the user element/restriction access

File: sdks/go/pkg/beam/core/runtime/exec/sdf.go
Comment (damccorm) [reply sdks/go/pkg/beam/core/runtime/exec/sdf.go:696]: I can't speak to why this decision was originally made (though I imagine it makes things significantly easier for the runner since input watermark is an important concept and := `min(all incoming output watermarks)`). It is worth noting that we're not actually setting _every_ output to this value, we're just setting the outputs of this transform to this value. At the runner level, I believe this is unioned with other output watermarks that may exist and this map is used to make triggering/windowing decisions

**Comment by damccorm** [comment]
R: @lostluck for a more watermark estimation focused review

File: sdks/go/pkg/beam/core/graph/fn.go
Comment (lostluck) [inline sdks/go/pkg/beam/core/graph/fn.go:983]: That seems like something we can make easier on ourselves (in a later PR), by adding a new function to our errors package so we don't need to repeat ourselves when we want a top level message, but also not lose that info when a higher level sets a top level error....

File: sdks/go/pkg/beam/core/runtime/exec/sdf.go
Comment (lostluck) [inline sdks/go/pkg/beam/core/runtime/exec/sdf.go:437]:
```suggestion
	// if one is defined, or nil otherwise.
```

File: sdks/go/pkg/beam/core/sdf/watermarkEstimator.go
Comment (lostluck) [inline sdks/go/pkg/beam/core/sdf/watermarkEstimator.go:20]: Please add documentation strings for exported types.

File: sdks/go/pkg/beam/core/sdf/watermarkEstimator.go
Comment (lostluck) [inline sdks/go/pkg/beam/core/sdf/watermarkEstimator.go:1]: WRT the file name, prefer snake_case over lowerCamelCase.

Or simplify the name to estimators.go or watermarks.go and avoid the question entirely.

File: sdks/go/pkg/beam/core/sdf/watermarkEstimator.go
Comment (lostluck) [inline sdks/go/pkg/beam/core/sdf/watermarkEstimator.go:20]: Exported types and methods on those types must have documentation comments.

**Review by lostluck — COMMENTED** [review]
I need to take another look at this to ensure I understand, but here are some comments to get started.

File: sdks/go/pkg/beam/core/graph/fn.go
Comment (damccorm) [reply sdks/go/pkg/beam/core/graph/fn.go:983]: Yeah, this file does that a lot - that's a good idea

File: sdks/go/pkg/beam/core/sdf/watermarkEstimator.go
Comment (damccorm) [reply sdks/go/pkg/beam/core/sdf/watermarkEstimator.go]: I went with snake case

File: sdks/go/pkg/beam/core/sdf/watermarkEstimator.go
Comment (damccorm) [reply sdks/go/pkg/beam/core/sdf/watermarkEstimator.go]: Done - good callout

Comment (damccorm) [reply sdks/go/pkg/beam/core/sdf/watermarkEstimator.go]: Done - I wonder if we can add tighter linting rules here?

File: sdks/go/pkg/beam/core/graph/fn.go
Comment (lostluck) [inline sdks/go/pkg/beam/core/graph/fn.go:958]:
```suggestion
	_, isWatermarkEstimating := fn.methods[createWatermarkEstimatorName]
```

File: sdks/go/pkg/beam/core/graph/fn.go
Comment (lostluck) [inline sdks/go/pkg/beam/core/graph/fn.go:960]:
```suggestion
		return false, errors.Errorf("watermark estimation method %v is defined on non-splittable DoFn. Watermark"+
```
https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: sdks/go/pkg/beam/core/runtime/exec/sdf.go
Comment (lostluck) [reply sdks/go/pkg/beam/core/runtime/exec/sdf.go:342]: For Daniel's benefit: In the contexts where Watermark estimation is being used, often only "1" element will be processed, since it's an extremely high fanout from that 1 element. So Bad for micro benchmarking, but probably fine in practice.

Ultimately, this feeds into my desire to have a single Very Good example of how to write a Streaming SDF with all the bells and whistles, so folks don't end up hurting themselves with all the new features we're adding.

**Review by lostluck — APPROVED** [review]
Two small style things, the internal linter would complain about, but otherwise LGTM.

**Comment by lostluck** [comment]
I think the difficulty here is that we can't meaningfully test the effectiveness of the watermark estimation until we have ProcessContinuations in, but I see (other than the last two suggested changes) no reason not to get it in at the moment.

**Comment by damccorm** [comment]
> I think the difficulty here is that we can't meaningfully test the effectiveness of the watermark estimation until we have ProcessContinuations in, but I see (other than the last two suggested changes) no reason not to get it in at the moment.

I would argue that we can to some degree with splitting and a little hacking to set the mode to streaming (I've run this with dataflow and seen it correctly update the watermarks on splits as described in my description). That isn't a complete demonstration of functionality, but the implementation isn't complete either since it doesn't try to handle processContinuations yet (that will need to be a future pr). Either way, I'm on board with your overall assessment that we shouldn't block this change on that, if for no other reason than it blocks future watermark work.

Thanks for the linting callouts, I committed both of them

**Pull request merged**
